### PR TITLE
GlobalKey docs improvement

### LINKED ISSUE #66625: Make it clear that GlobalKey should not be instantiated during the build phase
Instantiating a GlobalKey inside of the build method may cause some issues like https://github.com/flutter/flutter/issues/66427, we need to make it clear in the docs that GlobalKey should not be instantiated during the build phase. See https://github.com/flutter/flutter/issues/66427#issuecomment-697955513.

### PR DESCRIPTION
**Pull request by justinmc**
## Description

An [issue](https://github.com/flutter/flutter/issues/66427) was filed as an InteractiveViewer bug that ended up being just a misuse of GlobalKey.  It was noted in that issue that the docs should warn against instantiating GlobalKeys inside of build methods to avoid this in the future.  This PR adds this warning to GlobalKey's docs.

## Related Issues

https://github.com/flutter/flutter/issues/66427
Closes https://github.com/flutter/flutter/issues/66625

## Tests

None, docs change only.

### CODE REVIEW

**Comment by escamoteur** [comment]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -116,6 +116,19 @@ class ObjectKey extends LocalKey {
...
+/// so will create a new GlobalKey on every build, which will indicate that the
+/// corresponding widget subtree has changed on every build.
+///
+/// Besides harming performance, this can also cause unexpected behavior in
+/// widgets in the subtree. For example, a GestureDetector in the subtree will
+/// be unable to track ongoing gestures since it will be recreated on each
+/// build.
+///
+/// Instead, instantiate the GlobalKey outside of the build method, such as when
+/// the class instance is created.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:130]: maybe link to initState here?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -116,6 +116,19 @@ class ObjectKey extends LocalKey {
 /// You cannot simultaneously include two widgets in the tree with the same
 /// global key. Attempting to do so will assert at runtime.
 ///
+/// ## Pitfalls
+/// It is not advised to instantiate a GlobalKey inside of a build method. Doing
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:120]: maybe rephrase this to: GlobalKey are usually long-lived objects owned for example by a [State] object. They should not be re-created on every build.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -116,6 +116,19 @@ class ObjectKey extends LocalKey {
...
+/// so will create a new GlobalKey on every build, which will indicate that the
+/// corresponding widget subtree has changed on every build.
+///
+/// Besides harming performance, this can also cause unexpected behavior in
+/// widgets in the subtree. For example, a GestureDetector in the subtree will
+/// be unable to track ongoing gestures since it will be recreated on each
+/// build.
+///
+/// Instead, instantiate the GlobalKey outside of the build method, such as when
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:129]: If you follow this advise, you will still have the same problem for Stateless widgets that do get re-instantiated when they rebuild. You need to make sure the key is owned by soemthing longer lived, e.g. a state object.

**Comment by justinmc** [comment]
@goderbauer I rewrote it a bit with your suggestions, thanks for the feedback.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -116,6 +116,19 @@ class ObjectKey extends LocalKey {
...
 /// global key. Attempting to do so will assert at runtime.
 ///
+/// ## Pitfalls
+/// GlobalKeys should not be re-created on every build. They should usually be
+/// long-lived objects owned by a [State] object, for example.
+///
+/// Creating a new GlobalKey on every build will indicate to the framework that
+/// the corresponding widget subtree has changed on every build. Besides harming
+/// performance, this can also cause unexpected behavior in widgets in the
+/// subtree. For example, a GestureDetector in the subtree will be unable to
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:126]: `[GestureDetector]`

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -116,6 +116,19 @@ class ObjectKey extends LocalKey {
 /// You cannot simultaneously include two widgets in the tree with the same
 /// global key. Attempting to do so will assert at runtime.
 ///
+/// ## Pitfalls
+/// GlobalKeys should not be re-created on every build. They should usually be
+/// long-lived objects owned by a [State] object, for example.
+///
+/// Creating a new GlobalKey on every build will indicate to the framework that
+/// the corresponding widget subtree has changed on every build. Besides harming
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:124]: This sentence is kinda strange. Maybe something like: Creating a new GlobalKey on every build will throw away the state of the subtree associated with the old key and create a new fresh subtree for the new key.

**Comment by justinmc** [comment]
@goderbauer Done, thanks again!

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by pedromassangocode** [comment]
LGTM

**Review by pedromassangocode — APPROVED** [review]
LGTM
